docs/src/modules/ROOT/pages/enterprise-edition/enterprise-edition.adoc

```diff
@@ -716,5 +717,215 @@ the host is likely to hang or freeze,
+----
+
+Counterintuitively, the building blocks produced by these (seemly) identical methods are not necessarily functionality equivalent:
+
+[source,java,options="nowrap"]
+----
+UniConstraintStream<Shift> a(ConstraintFactory constraintFactory) {
+    return factory.forEach(Shift.class)
+                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
+}
+
+UniConstraintStream<Shift> b(ConstraintFactory constraintFactory) {
+    return factory.forEach(Shift.class)
+                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
+}
+----
+
+The Java Virtual Machine is free to (and often does) create different instances of functionality equivalent lambdas.
+This severely limits the effectiveness of node sharing, since the only way to know two lambdas are equal is to compare their references.
+
+When automatic node sharing is used, the `ConstraintProvider` class is transformed so all lambdas are accessed via a static final field.
+So a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
 
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+is transformed to a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+    private static final Predicate<Shift> $predicate1 = shift -> shift.getEmployee().getName().equals("Ann");
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
+
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter($predicate1)
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+IMPORTANT: This transformation prevents breakpoints from being placed inside the ConstraintProvider.
```

Comment:
As explain above, this needs to be up there. But because it can not mention this transformation, it needs to be restated. Something like "Debugging breakpoints put inside your constraints will not be respected, because the `ConstraintProvider` class will be transformed when this feature is enabled".

Then when you're back down here, you can mention it again - and now they will understand _why_. Those that will care. Which, again, most won't.